termux-exec.c

```diff
@@ -8,231 +15,508 @@
+
+  result->interpreter =
+      termux_rewrite_executable(interpreter, result->interpreter_buf, sizeof(result->interpreter_buf));
 }
+
+// Interceptor of the execve(2) system call using LD_PRELOAD.
+int execve(const char *executable_path, char *const argv[], char *const envp[]) {
+  const bool termux_exec_debug = getenv("TERMUX_EXEC_DEBUG") != NULL;
+  if (termux_exec_debug) {
+    fprintf(stderr, "[termux-exec] Intercepting execve('%s'):\n", executable_path);
+    int tmp_argv_count = 0;
+    while (argv[tmp_argv_count] != NULL) {
+      fprintf(stderr, "[termux-exec]    arg[%d] = %s\n", tmp_argv_count, argv[tmp_argv_count]);
+      tmp_argv_count++;
+    }
+  }
+
+  const char *orig_executable_path = executable_path;
+
+  char executable_path_buffer[PATH_MAX];
+  executable_path = termux_rewrite_executable(executable_path, executable_path_buffer, sizeof(executable_path_buffer));
+  if (termux_exec_debug) {
+    fprintf(stderr, "[termux-exec] Possibly rewritten path: '%s'\n", executable_path);
+  }
+
+  if (access(executable_path, X_OK) != 0) {
+    // Error out if the file is not executable:
+    errno = EACCES;
+    return -1;
+  }
+
+  int fd = open(executable_path, O_RDONLY);
+  if (fd == -1) {
+    errno = ENOENT;
+    return -1;
+  }
+
+  // execve(2): "The kernel imposes a maximum length on the text that follows the "#!" characters
+  // at the start of a script; characters beyond the limit are ignored. Before Linux 5.1, the
+  // limit is 127 characters. Since Linux 5.1, the limit is 255 characters."
+  // We use one more byte since inspect_file_header() will null terminate the buffer.
+  char header[256];
+  ssize_t read_bytes = read(fd, header, sizeof(header) - 1);
+  close(fd);
+
+  struct file_header_info info = {
+      .interpreter = NULL,
+      .interpreter_arg = NULL,
+  };
+  inspect_file_header(header, read_bytes, &info);
+
+  if (!info.is_elf && info.interpreter == NULL) {
+    errno = ENOEXEC;
+    return -1;
+  }
+
+  if (info.interpreter != NULL) {
+    executable_path = info.interpreter;
+  }
+
+  // Resolve the absolute path of the binary to execute so we can tell if.
+  char realpath_buffer[PATH_MAX];
+  bool realpath_ok = realpath(executable_path, realpath_buffer) != NULL;
+
+  char normalized_path_buffer[PATH_MAX];
+
+  char **new_allocated_envp = NULL;
+
+  const char **new_argv = NULL;
+  if (realpath_ok) {
+    bool wrap_in_linker =
+#ifdef __ANDROID__
+        android_get_device_api_level() >= 29 &&
+#endif
+        (strstr(realpath_buffer, TERMUX_BASE_DIR) != NULL);
+
+    if (termux_exec_debug) {
+      fprintf(stderr, "[termux-exec] realpath(\"%s\") = \"%s\", wrapping=%s\n", executable_path, realpath_buffer,
+              wrap_in_linker ? "yes" : "no");
+    }
+
+    bool cleanup_env = info.is_non_native_elf ||
+                       (starts_with(executable_path, "/system/") && !starts_with(executable_path, "/system/bin/sh"));
+    if (cleanup_env) {
+      remove_ld_from_env(envp, &new_allocated_envp);
```

Comment:
hmmm, this may cause problems for use cases. Some users export them, also used by `sudo`/`tudo`.

Reply:
Note that this behaviour is already in place in `termux-exec` (https://github.com/termux/termux-exec/blob/master/termux-exec.c), although slightly changed and reorganised here.

Reply:
Ah, right. Will have to be checked.

https://github.com/agnostic-apollo/sudo#path-and-ld_library_path-priorities

Reply:
Actually, shouldn't be an issue for `sudo`/`tudo`, they unset `LD_PRELOAD` for system binaries, had forgotten about it, been long.

https://github.com/agnostic-apollo/sudo/blob/743770c88680c1dbc00a0ade7840fab8c48ad0e2/sudo#L1246C1-L1246C2